cmd/sink/main.go

```diff
@@ -5,22 +5,151 @@ package main
 
 import (
 	"context"
+	"database/sql"
 	"log"
 	"net/http"
 	"os"
+	"time"
 
 	ceOtelObs "github.com/cloudevents/sdk-go/observability/opentelemetry/v2/client"
 	cloudevents "github.com/cloudevents/sdk-go/v2"
 	ceClient "github.com/cloudevents/sdk-go/v2/client"
+	"github.com/cloudevents/sdk-go/v2/types"
 	kaObservability "github.com/kubearchive/kubearchive/pkg/observability"
+	_ "github.com/lib/pq"
 	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
 )
 
-var logger = log.New(os.Stderr, "", log.LstdFlags|log.Lmicroseconds|log.LUTC)
+const (
+	connStr             = "user=kubearchive password=kubearchive dbname=kubearchive-db host=10.96.22.175 port=5432 sslmode=disable"
+	apiVersionExtension = "apiversion"
+	kindExtension       = "kind"
+	nameExtension       = "name"
+	namespaceExtension  = "namespace"
+)
+
+var (
+	logger         = log.New(os.Stderr, "", log.LstdFlags|log.Lmicroseconds|log.LUTC)
+	db     *sql.DB = nil
+)
+
+// Represents the fields in a resource's metadata that are need to write a resource into the database
+type ResourceMetadata struct {
+	ResourceVersion string `json:"resourceVersion"`
+}
+
+// Respresents the fields in the data field of a cloudevent to write a resource into the database
+type ResourceData struct {
+	Created     string           `json:"firstTimestamp"`
+	LastUpdated string           `json:"lastTimestamp"`
+	Metadata    ResourceMetadata `json:"metadata"`
+}
+
+// checks that the cloudevents has the appropriate extensions set and have values that are the right type. Additionally
+// it checks that the cloudevent's data has the necessary fields. If all conditions are not met, it returns an error
+func validateCloudevent(event cloudevents.Event) error {
+	eventExtensions := event.Extensions()
+	_, err := types.ToString(eventExtensions[apiVersionExtension])
+	if err != nil {
+		return err
+	}
+	_, err = types.ToString(eventExtensions[kindExtension])
+	if err != nil {
+		return err
+	}
+	_, err = types.ToString(eventExtensions[nameExtension])
+	if err != nil {
+		return err
+	}
+	_, err = types.ToString(eventExtensions[namespaceExtension])
+	if err != nil {
+		return err
+	}
+	payload := ResourceMetadata{}
+	err = event.DataAs(&payload)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Writes the kubernetes resource to the database. This method must be called after calling validateCloudevent and it
+// does not return an error
+func writeToDatabase(event cloudevents.Event) {
+	var err error
+	if db == nil {
+		logger.Println("opening connection to the database")
+		db, err = sql.Open("postgres", connStr)
+		if err != nil {
+			logger.Printf("failed to connect to database: %s", err)
+			return
+		}
+	}
 
+	eventExtensions := event.Extensions()
+	// since a call to validateCloudevent should already have passed we can assume that all of these values exist as
+	// strings
+	apiVersion, _ := eventExtensions[apiVersionExtension]
```

Comment:
To fix this so you don't get the security warning:

```suggestion
	apiVersion := eventExtensions[apiVersionExtension]
```

Reply:
done